### PR TITLE
dev-embedded/sdcc: Bump to 4.2.0

### PR DESCRIPTION
**Pull request by mid-kid**
This ebuild fixes all long-standing bugs I could find on the tracker, as I've been able to confirm myself that PIC projects build correctly. Other targets haven't been tested but build fine.

I'm not going to backport the fixes to the previous ebuilds, as I'm not entirely sure they would apply correctly. Similarly, -9999 is untested.

This would also obsolete the following PR: https://github.com/gentoo/gentoo/pull/16552

### CODE REVIEW

**Comment by mid-kid** [comment]
holdon, getting segfaults with this new gpasm...

**Comment by mid-kid** [comment]
Only with LTO enabled, apparently. Otherwise it works fine. Trying to debug, but might be a pain.

**Comment by mid-kid** [comment]
filter-flags'd lto for the time being. each sloppy fix I made revealed three other bugs, not sitting through that shit.

File: dev-embedded/sdcc/sdcc-4.2.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+"
+
+for port in ${SDCC_PORTS}; do
+REQUIRED_USE="${REQUIRED_USE}
+	${port}? ( sdbinutils )
+"
+done
+REQUIRED_USE="${REQUIRED_USE}
+	|| ( ${SDCC_PORTS} )
+"
```
Comment (ceamac) [inline dev-embedded/sdcc/sdcc-4.2.0.ebuild:58]: `REQUIRED_USE` requires at least one port, and any port requires `sdbinutils`, so this flag is always on.
I removed it and set it to always enabled in a follow-up commit.

This fixed some pkgcheck warnings.

File: dev-embedded/sdcc/sdcc-4.2.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+	sdbinutils? ( sys-libs/zlib:= )
+	sdcdb? ( sys-libs/readline:0= )
+	ucsim? ( sys-libs/ncurses:= )
+"
+DEPEND="
+	${RDEPEND}
+	dev-util/gperf
+"
+PATCHES=(
+	"${FILESDIR}"/sdcc-3.8.0-override-override.patch
```
Comment (ceamac) [inline dev-embedded/sdcc/sdcc-4.2.0.ebuild:74]: This does not apply cleanly in the live ebuild, so it cannot be built with automake 2.71.

Will not fix it now, but let's keep it in mind.

File: dev-embedded/sdcc/sdcc-4.2.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	use pic16 && libs+=( device/lib/build/pic16/libsdcc.lib )
+	for lib in "${libs[@]}"; do
+		[[ -f "${lib}" ]] || die "Failed to build ${lib}"
+	done
+}
+
+src_install() {
+	default
+	dodoc doc/*.txt
+	find "${D}" -type d -name .deps -exec rm -vrf {} + || die
```
Comment (ceamac) [inline dev-embedded/sdcc/sdcc-4.2.0.ebuild:161]: `${D}` -> `${ED}`

WIll fix it while merging

**Review by ceamac — APPROVED** [review]
Great work!  Thanks!

**Comment by mid-kid** [comment]
Thanks a lot for fixing it up!

**Pull request closed without merging**